pkg/addonmanager/cloudevents/manager.go

```diff
@@ -0,0 +1,348 @@
+
+	kubeInformers := kubeinformers.NewSharedInformerFactoryWithOptions(kubeClient, 10*time.Minute,
+		kubeinformers.WithTweakListOptions(func(listOptions *metav1.ListOptions) {
+			selector := &metav1.LabelSelector{
+				MatchExpressions: []metav1.LabelSelectorRequirement{
+					{
+						Key:      addonv1alpha1.AddonLabelKey,
+						Operator: metav1.LabelSelectorOpIn,
+						Values:   addonNames,
+					},
+				},
+			}
+			listOptions.LabelSelector = metav1.FormatLabelSelector(selector)
+		}),
+	)
+
+	workClient := clientHolder.WorkInterface()
+	workInformers := clientHolder.ManifestWorkInformer()
+
+	// addonDeployController
+	err = workInformers.Informer().AddIndexers(
+		cache.Indexers{
+			index.ManifestWorkByAddon:           index.IndexManifestWorkByAddon,
+			index.ManifestWorkByHostedAddon:     index.IndexManifestWorkByHostedAddon,
+			index.ManifestWorkHookByHostedAddon: index.IndexManifestWorkHookByHostedAddon,
+		},
+	)
+	if err != nil {
+		return err
+	}
+
+	err = addonInformers.Addon().V1alpha1().ManagedClusterAddOns().Informer().AddIndexers(
+		cache.Indexers{
+			index.ManagedClusterAddonByNamespace: index.IndexManagedClusterAddonByNamespace, // addonDeployController
+			index.ManagedClusterAddonByName:      index.IndexManagedClusterAddonByName,      // addonConfigController
+			index.AddonByConfig:                  index.IndexAddonByConfig,                  // addonConfigController
+		},
+	)
+	if err != nil {
+		return err
+	}
+
+	err = addonInformers.Addon().V1alpha1().ClusterManagementAddOns().Informer().AddIndexers(
+		cache.Indexers{
+			index.ClusterManagementAddonByConfig:    index.IndexClusterManagementAddonByConfig,    // managementAddonConfigController
+			index.ClusterManagementAddonByPlacement: index.IndexClusterManagementAddonByPlacement, // addonConfigController
+		})
+	if err != nil {
+		return err
+	}
+
+	err = a.startWithInformers(ctx, workClient, workInformers, kubeInformers, addonInformers, clusterInformers, dynamicInformers)
+	if err != nil {
+		return err
+	}
+
+	kubeInformers.Start(ctx.Done())
+	go workInformers.Informer().Run(ctx.Done())
+	addonInformers.Start(ctx.Done())
+	clusterInformers.Start(ctx.Done())
+	dynamicInformers.Start(ctx.Done())
+	return nil
+}
+
+func (a *cloudeventsAddonManager) StartWithInformers(ctx context.Context,
+	kubeInformers kubeinformers.SharedInformerFactory,
+	workInformers workinformers.SharedInformerFactory,
+	addonInformers addoninformers.SharedInformerFactory,
+	clusterInformers clusterv1informers.SharedInformerFactory,
+	dynamicInformers dynamicinformer.DynamicSharedInformerFactory) error {
+	// TODO: Implement this method to support usage within the addon template.
+	// Requires updating sdk-go to enable clientHolder to manage the shared informer for ManifestWork.
+	return fmt.Errorf("method StartWithInformers is not implemented")
+}
+
+func (a *cloudeventsAddonManager) startWithInformers(ctx context.Context,
```

Comment:
actully, how about we have a baseManager interface with StartWithInformers method and an implementation that new all controllers. The Manager interface inlines the baseManager interface, and the manager/cloudeventmanger have different implementation of start func, but share the same implementation in baseManager. Will that work?

Reply:
Updated this, not sure if this is more clear.

Reply:
why we still need this? I thought this should be in the baseManager.

Reply:
removed this.

Reply:
I mean all the code to init controllers and start them. I think they are duplicated code and should exist only in baseManager. If we have any change on controllers in the future, we do not need to update multiple times.

Reply:
We may not do that, due to the limitation source work client from sdk-go as I mentioned at https://github.com/open-cluster-management-io/addon-framework/pull/256#discussion_r1566694777

we can't get the `workinformers.SharedInformerFactory` from `clientHolder` (from sdk-go), we can only get `workv1informers.Informer` type, so we can't call `StartWithInformers` to start the controllers in cloudevents addon manager.